File: prediction-python/app/core/market_hours.py
"""Market-hours awareness (docs/CONTRACTS.md Addendum 1).

Pure functions, no I/O.  Two calendars:

* IR_GOLD_18K trades 24h/day on Iranian trading days: its primary source
  (Milli Gold, milli.gold) is an online platform with no intraday session —
  but the Iranian off-days still apply: closed all Thursday and Friday
  (Tehran).  During open days a paused feed goes honestly stale.
* Other Iranian symbols (USD_IRT, IR_COIN_EMAMI): open Sat-Wed between
  ``MARKET_TEHRAN_OPEN`` and ``MARKET_TEHRAN_CLOSE`` (Asia/Tehran local time,
  a fixed UTC+03:30 since Iran abolished DST); closed all Thursday and Friday.
* Global symbols (XAUUSD, XAGUSD, BRENT_OIL, DXY, US10Y): closed from
  Friday 21:00 UTC to Sunday 22:00 UTC, open otherwise.

Freshness rule while a market is CLOSED: an observation from the last session
(observed no earlier than ``closure start - STALE_MINUTES``) still counts as
acceptably fresh; anything older is truly stale.  While OPEN the plain
``STALE_MINUTES`` age rule applies unchanged.

Symbols with no known calendar are treated as always open (plain age rule).
"""
from __future__ import annotations

from datetime import datetime, time, timedelta, timezone
from typing import Optional
from zoneinfo import ZoneInfo

from ..config import Settings

TEHRAN = ZoneInfo("Asia/Tehran")
THURSDAY = 3  # Python weekday(): Monday=0 .. Sunday=6
FRIDAY = 4

# 24h-traded on Iranian trading days (primary source: Milli Gold, no intraday
# session window; Thursday+Friday off-days still apply).
IRAN_24H_SYMBOLS = frozenset({"IR_GOLD_18K"})
IRANIAN_SYMBOLS = frozenset({"USD_IRT", "IR_COIN_EMAMI"})
GLOBAL_SYMBOLS = frozenset({"XAUUSD", "XAGUSD", "BRENT_OIL", "DXY", "US10Y"})
# Tehran-exchange gold funds trade Sat-Wed between MARKET_TSE_OPEN and
# MARKET_TSE_CLOSE (default 12:00-17:00 Tehran); closed Thursday AND Friday.
TSE_FUND_PREFIX = "IR_GOLD_FUND"

GLOBAL_CLOSE_UTC = time(21, 0)  # Friday
GLOBAL_OPEN_UTC = time(22, 0)   # Sunday


def _parse_hhmm(raw: str, default: time) -> time:
    """'HH:MM' -> datetime.time, falling back to ``default`` on bad input."""
    try:
        hh, mm = str(raw).strip().split(":")
        return time(int(hh), int(mm))
    except (TypeError, ValueError):
        return default


def _ensure_utc(dt: datetime) -> datetime:
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def is_market_open(symbol: str, at_utc: datetime, settings: Settings) -> bool:
    """True when ``symbol``'s market is open at ``at_utc`` (aware or naive-UTC)."""
    at_utc = _ensure_utc(at_utc)
    if symbol in IRAN_24H_SYMBOLS:
        return at_utc.astimezone(TEHRAN).weekday() not in (THURSDAY, FRIDAY)
    if symbol.startswith(TSE_FUND_PREFIX):
        local = at_utc.astimezone(TEHRAN)
        if local.weekday() in (THURSDAY, FRIDAY):
            return False
        open_t = _parse_hhmm(getattr(settings, "market_tse_open", "12:00"), time(12, 0))
        close_t = _parse_hhmm(getattr(settings, "market_tse_close", "17:00"), time(17, 0))
        return open_t <= local.time() < close_t
    if symbol in IRANIAN_SYMBOLS:
        local = at_utc.astimezone(TEHRAN)
        if local.weekday() in (THURSDAY, FRIDAY):
            return False
        open_t = _parse_hhmm(settings.market_tehran_open, time(12, 0))
        close_t = _parse_hhmm(settings.market_tehran_close, time(20, 0))
        return open_t <= local.time() < close_t
    if symbol in GLOBAL_SYMBOLS:
        wd = at_utc.weekday()
        if wd == FRIDAY and at_utc.time() >= GLOBAL_CLOSE_UTC:
            return False
        if wd == 5:  # Saturday: closed all day
            return False
        if wd == 6 and at_utc.time() < GLOBAL_OPEN_UTC:
            return False
        return True
    return True  # unknown symbol: no calendar, treated as always open


def closure_started_at(
    symbol: str, at_utc: datetime, settings: Settings
) -> Optional[datetime]:
    """UTC start of the closure containing ``at_utc``; None while open.

    For 18k (24h symbol) that is Thursday 00:00 Tehran of the current
    Thu+Fri block; for windowed Iranian symbols the most recent trading-day
    close (``MARKET_TEHRAN_CLOSE`` on the latest Sat-Wed day at or before
    now); for global symbols the most recent Friday 21:00 UTC.
    """
    at_utc = _ensure_utc(at_utc)
    if is_market_open(symbol, at_utc, settings):
        return None
    if symbol in IRAN_24H_SYMBOLS:
        # Closed only during the Thu+Fri block; closure began at
        # Thursday 00:00 Tehran of the current block.
        local = at_utc.astimezone(TEHRAN)
        day = local.date()
        if day.weekday() == FRIDAY:
            day -= timedelta(days=1)
        return datetime.combine(day, time(0, 0), tzinfo=TEHRAN).astimezone(timezone.utc)
    if symbol.startswith(TSE_FUND_PREFIX):
        close_t = _parse_hhmm(getattr(settings, "market_tse_close", "17:00"), time(17, 0))
        local = at_utc.astimezone(TEHRAN)
        for days_back in range(9):
            day = (local - timedelta(days=days_back)).date()
            if day.weekday() in (THURSDAY, FRIDAY):
                continue  # no session, hence no close
            candidate = datetime.combine(day, close_t, tzinfo=TEHRAN)
            if candidate <= local:
                return candidate.astimezone(timezone.utc)
        return at_utc  # unreachable with a sane open<close configuration
    if symbol in IRANIAN_SYMBOLS:
        close_t = _parse_hhmm(settings.market_tehran_close, time(20, 0))
        local = at_utc.astimezone(TEHRAN)
        for days_back in range(9):
            day = (local - timedelta(days=days_back)).date()
            if day.weekday() in (THURSDAY, FRIDAY):
                continue  # off-days never have a session, hence no close
            candidate = datetime.combine(day, close_t, tzinfo=TEHRAN)
            if candidate <= local:
                return candidate.astimezone(timezone.utc)
        return at_utc  # unreachable with a sane open<close configuration
    if symbol in GLOBAL_SYMBOLS:
        days_back = (at_utc.weekday() - FRIDAY) % 7
        candidate = datetime.combine(
            (at_utc - timedelta(days=days_back)).date(),
            GLOBAL_CLOSE_UTC,
            tzinfo=timezone.utc,
        )
        if candidate > at_utc:
            candidate -= timedelta(days=7)
        return candidate
    return None


def is_acceptably_fresh(
    symbol: str,
    observed_at: Optional[datetime],
    at_utc: datetime,
    settings: Settings,
) -> bool:
    """Market-hours-aware staleness check (Addendum 1).

    * market OPEN   -> age <= STALE_MINUTES (unchanged semantics);
    * market CLOSED -> ``observed_at`` must be no older than
      ``closure start - STALE_MINUTES``, i.e. data from the last session
      keeps counting as fresh for the whole closure.
    """
    if observed_at is None:
        return False
    observed_at = _ensure_utc(observed_at)
    at_utc = _ensure_utc(at_utc)
    tolerance = timedelta(minutes=settings.stale_minutes)
    closure_start = closure_started_at(symbol, at_utc, settings)
    if closure_start is None:  # market open: plain age rule
        return (at_utc - observed_at) <= tolerance
    return observed_at >= closure_start - tolerance

File: backend-go/internal/markethours/markethours.go
// Package markethours implements the Addendum 1 market-calendar rules,
// mirrored between the Go and Python services from the same env vars:
//
//   - IR_GOLD_18K trades 24h/day on Iranian trading days: its primary source
//     (Milli Gold) is an online platform with no intraday session, but the
//     Iranian off-days still apply — closed all Thursday and Friday (Tehran).
//   - Other Iranian symbols (USD_IRT, IR_COIN_EMAMI) trade Sat-Wed between
//     MARKET_TEHRAN_OPEN and MARKET_TEHRAN_CLOSE (Asia/Tehran local,
//     open inclusive, close exclusive); closed all Thursday and Friday.
//   - Global symbols (XAUUSD, XAGUSD, BRENT_OIL, DXY, US10Y, ...) are closed
//     from Friday 21:00 UTC until Sunday 22:00 UTC.
//
// Freshness follows the addendum: while a market is OPEN, data older than
// STALE_MINUTES is stale; while CLOSED, data observed during the last session
// (observed_at >= closure start - STALE_MINUTES) is still acceptably fresh.
package markethours

import (
	"strings"
	"time"
)

// Contract defaults for the Tehran session (Asia/Tehran local, HH:MM).
const (
	DefaultOpen  = "12:00"
	DefaultClose = "20:00"
)

// TSE gold-fund session (Addendum 7): Sat-Wed 12:00-18:00 Asia/Tehran,
// closed Thursday AND Friday. Fixed here (display freshness only); the
// Python service reads MARKET_TSE_OPEN/CLOSE for prediction-side rules.
const (
	tseOpen  = "12:00"
	tseClose = "18:00"
)

// usdOpen: the free-market USD session opens earlier than the gold bazaar
// (close is shared). Fixed here for display freshness; the Python service
// reads MARKET_USD_OPEN for prediction-side rules.
const usdOpen = "10:00"

// tseFundPrefix marks Tehran-exchange gold-fund symbols.
const tseFundPrefix = "IR_GOLD_FUND"

// iran24h: symbols whose primary source trades around the clock on Iranian
// trading days. IR_GOLD_18K's primary provider is Milli Gold (milli.gold),
// a 24-hour online platform — no intraday session window, but the Iranian
// off-days (Thursday + Friday, Tehran) still close the market.
var iran24h = map[string]bool{
	"IR_GOLD_18K": true,
}

// iranian is the set of symbols that follow the Tehran bazaar calendar.
// Every other symbol follows the global (UTC weekend) calendar.
var iranian = map[string]bool{
	"USD_IRT":       true,
	"IR_COIN_EMAMI": true,
}

// tehran is the Asia/Tehran location. The runtime container installs tzdata
// (apk) and Go toolchains ship zoneinfo, so LoadLocation normally succeeds;
// the fallback is the fixed +03:30 offset (Iran abolished DST in 2022).
var tehran = loadTehran()

func loadTehran() *time.Location {
	if loc, err := time.LoadLocation("Asia/Tehran"); err == nil {
		return loc
	}
	return time.FixedZone("Asia/Tehran", 3*3600+30*60)
}

// parseHHMM parses "HH:MM" into minutes since midnight, falling back to the
// given default when the value is empty or malformed (config validates the
// env vars at boot; the fallback keeps this package total).
func parseHHMM(s, def string) int {
	t, err := time.Parse("15:04", s)
	if err != nil {
		t, _ = time.Parse("15:04", def)
	}
	return t.Hour()*60 + t.Minute()
}

// IsOpen reports whether the market for symbol is open at the instant `at`.
// open/close are "HH:MM" Tehran-local session bounds (only used for Iranian
// symbols); pass the configured MARKET_TEHRAN_OPEN / MARKET_TEHRAN_CLOSE.
func IsOpen(symbol string, at time.Time, open, close string) bool {
	if iran24h[symbol] {
		wd := at.In(tehran).Weekday()
		return wd != time.Thursday && wd != time.Friday
	}
	if strings.HasPrefix(symbol, tseFundPrefix) {
		lt := at.In(tehran)
		if lt.Weekday() == time.Thursday || lt.Weekday() == time.Friday {
			return false
		}
		m := lt.Hour()*60 + lt.Minute()
		return m >= parseHHMM(tseOpen, tseOpen) && m < parseHHMM(tseClose, tseClose)
	}
	if iranian[symbol] {
		lt := at.In(tehran)
		if lt.Weekday() == time.Thursday || lt.Weekday() == time.Friday {
			return false
		}
		m := lt.Hour()*60 + lt.Minute()
		openM := parseHHMM(open, DefaultOpen)
		if symbol == "USD_IRT" {
			openM = parseHHMM(usdOpen, usdOpen)
		}
		return m >= openM && m < parseHHMM(close, DefaultClose)
	}
	u := at.UTC()
	switch u.Weekday() {
	case time.Friday:
		return u.Hour() < 21
	case time.Saturday:
		return false
	case time.Sunday:
		return u.Hour() >= 22
	default:
		return true
	}
}

// ClosureStartedAt returns the UTC instant the closure containing `at`
// began: the end of the most recent trading session. When the market is open
// at `at` it returns `at` itself (no closure in progress).
func ClosureStartedAt(symbol string, at time.Time, open, close string) time.Time {
	if IsOpen(symbol, at, open, close) {
		return at.UTC()
	}
	if iran24h[symbol] {
		// Closed only during the Thu+Fri block; the closure began at
		// Thursday 00:00 Tehran of the current block.
		lt := at.In(tehran)
		day := time.Date(lt.Year(), lt.Month(), lt.Day(), 0, 0, 0, 0, tehran)
		if lt.Weekday() == time.Friday {
			day = day.AddDate(0, 0, -1)
		}
		return day.UTC()
	}
	if strings.HasPrefix(symbol, tseFundPrefix) {
		closeM := parseHHMM(tseClose, tseClose)
		lt := at.In(tehran)
		day := time.Date(lt.Year(), lt.Month(), lt.Day(), 0, 0, 0, 0, tehran)
		for i := 0; i < 9; i++ {
			d := day.AddDate(0, 0, -i)
			if d.Weekday() == time.Thursday || d.Weekday() == time.Friday {
				continue
			}
			closeT := d.Add(time.Duration(closeM) * time.Minute)
			if !closeT.After(lt) {
				return closeT.UTC()
			}
		}
		return at.UTC() // unreachable
	}
	if iranian[symbol] {
		closeM := parseHHMM(close, DefaultClose)
		lt := at.In(tehran)
		day := time.Date(lt.Year(), lt.Month(), lt.Day(), 0, 0, 0, 0, tehran)
		// Walk back to the most recent trading-day close at or before `at`.
		// Any 9-day span contains a non-Thu/Fri close in the past.
		for i := 0; i < 9; i++ {
			d := day.AddDate(0, 0, -i)
			if d.Weekday() == time.Thursday || d.Weekday() == time.Friday {
				continue
			}
			closeT := d.Add(time.Duration(closeM) * time.Minute)
			if !closeT.After(lt) {
				return closeT.UTC()
			}
		}
		return at.UTC() // unreachable
	}
	// Global markets close Friday 21:00 UTC.
	u := at.UTC()
	daysBack := (int(u.Weekday()) - int(time.Friday) + 7) % 7
	fri := time.Date(u.Year(), u.Month(), u.Day(), 21, 0, 0, 0, time.UTC).AddDate(0, 0, -daysBack)
	if fri.After(u) {
		fri = fri.AddDate(0, 0, -7)
	}
	return fri
}

// AcceptablyFresh reports whether an observation from observedAt is
// acceptably fresh at `at` under the Addendum 1 rules:
//
//	market open:   age <= staleMinutes
//	market closed: observedAt >= closure start - staleMinutes
//	               (i.e. last-session data never goes stale overnight)
func AcceptablyFresh(symbol string, observedAt, at time.Time, staleMinutes int, open, close string) bool {
	stale := time.Duration(staleMinutes) * time.Minute
	if IsOpen(symbol, at, open, close) {
		return at.Sub(observedAt) <= stale
	}
	return !observedAt.Before(ClosureStartedAt(symbol, at, open, close).Add(-stale))
}
